### PR TITLE
Network document

### PR DESCRIPTION
**Pull request by yaocw2020**
Add document to explain the principles of Harvester networking, including management network and VLAN network.

### CODE REVIEW

**Comment by guangbochen** [comment]
@yaocw2020 can u resolve the conflict and do not include any code changes of the doc PR.

**Comment by yaocw2020** [comment]
> @yaocw2020 can u resolve the conflict and do not include any code changes of the doc PR.

Done

File: docs/harvester-network.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  ![image-20210408235822223](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210408235822223.png)
+
+  If there is no error, click **Create** Button to create a VLAN network.
+
+  ![image-20210408232931111](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210408232931111-20210409000151223.png)
+
+- When creating or editing VM, select the **Networks** tab, you can change the existing network or add new network with the VLAN network created before. 
+
+  ![image-20210408233240281](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210408233240281.png)
```
Comment (guangbochen) [inline docs/harvester-network.md]: - Enable the default VLAN network configuration via ...
- Input a valid default physical NIC name of the VLAN network, the first physical NIC name of each Harvester node always defaults to `eth0`, it is recommended to choose a separate NIC for the VLAN other than the one used for the management network(the one selected during the Harvester installation) for better network performance and isolation. (Note: modify the default VLAN network setting will not change the existing configured host networks)
- optional) users can always customize each node's VLAN network configuration via ...
- create a new VLAN network via ... (Note: You can use VLAN 1 for simple installations where there is no VLAN trunking, but good practice is to not put user data on VLAN 1 in a multi-VLAN environment)
- Creating a VM and add the network configurations
    - only the first network card will be enabled by default, the user can either choose to use a management network or a VLAN network. (Note: will need to select `Install guest agent` option in the `Advanced Options` tab to get the VLAN network IP address from the Harvester UI)
 image
    - users can choose to add one or multiple network cards, the additional network card configurations can be set via cloud-init network data. e.g.,
```YAML
version: 1
config:
  - type: physical
    name: enp1s0 # name is various upon OS image
    subnets:
      - type: dhcp
  - type: physical
    name: enp2s0 
    subnets:
      - type: DHCP
```

File: docs/harvester-network.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+## Principles
+### Management network
+
+Harvester  adopts [flannel](https://github.com/flannel-io/flannel)  as the default CNI to implement management network. It's an internal network, exactly, that means it should be within cluster if you want to access a VM via the management network.
+
+### VLAN
+
+[Harvester network controller](https://github.com/rancher/harvester-network-controller) provides an VLAN network implementation for Harvester with the help of [multus](https://github.com/k8snetworkplumbingwg/multus-cni) and [bridge CNI](https://www.cni.dev/plugins/current/main/bridge/).
+
+Here is the VLAN network principle diagram.
```
[the remark this answers is not in the document]
Comment (yaocw2020) [reply docs/harvester-network.md]: > You can use VLAN 1 for simple installations where there is no VLAN trunking, but good practice is to not put user data on VLAN 1 in a multi-VLAN environment

This statement is not accurate.  There should be an additional condition, that is, the PVID is 1.

File: docs/harvester-network.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+#### Management Network
+
+Harvester adopts [flannel](https://github.com/flannel-io/flannel) as the default CNI to implement the management network. It's an internal network, which means the user can only access the VM's management network within its cluster nodes or pods.
+
+#### VLAN
+
+[Harvester network controller](https://github.com/rancher/harvester-network-controller) leverages the [multus](https://github.com/k8snetworkplumbingwg/multus-cni) and [bridge](https://www.cni.dev/plugins/current/main/bridge/) CNI plugins to implement the VLAN.  
+
+Below is a user case of the VLAN in Harvester.
```
Comment (guangbochen) [inline docs/harvester-network.md]: \s\user\use

File: docs/harvester-network.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+* [Introduction of The Harvester network](#introduction-of-the-harvester-network)
+  * [Summary](#summary)
+  * [Implementation](#implementation)
+    * [Management Network](#management-network)
+    * [VLAN](#vlan)
+  * [UI Interaction](#ui-interaction)
+
+
+### Summary
+[Harvester](https://github.com/rancher/harvester) is built on Kubernetes, which uses [CNI](https://github.com/containernetworking/cni) as an interface between network providers and Kubernetes pod networking. Naturally, we implement the Harvester network based on CNI. Moreover, [Harvester UI](https://github.com/rancher/harvester-ui) integrates the Harvester network to provide a user-friendly way to configure network for VMs.
```
Comment (guangbochen) [inline docs/harvester-network.md]: \s\configure network\configure the network

File: docs/harvester-network.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+#### VLAN
+
+[Harvester network controller](https://github.com/rancher/harvester-network-controller) leverages the [multus](https://github.com/k8snetworkplumbingwg/multus-cni) and [bridge](https://www.cni.dev/plugins/current/main/bridge/) CNI plugins to implement the VLAN.  
+
+Below is a user case of the VLAN in Harvester.
+
+![image-20210412181210146](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210412181210146.png)
+
+- Harvester network controller use a bridge for a node and a pair of veth for a VM to implement the VLAN. The bridge acts as a switch to forward the network traffic from or to VMs and the veth pair is like the connected ports between vms and switch.
```
Comment (guangbochen) [inline docs/harvester-network.md]: \s\network controller\network-controller uses

File: docs/harvester-network.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+- VMs within the same VLAN is able to communicate with each other, while the VMs within different VLANs cann't.
```
Comment (guangbochen) [inline docs/harvester-network.md]: \s\cann't\can't

File: docs/harvester-network.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+- The external switch ports conneted with the hosts or other devices(such as DHCP server) should be set as trunk or hybrid type and permit the specified VLANs.
```
Comment (guangbochen) [inline docs/harvester-network.md]: \s\connect\connected

File: docs/harvester-network.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+### UI Interaction
+
+- Enable VLAN via going to **Setting > vlan** to enable VLAN and input a valid default physical NIC name for the VLAN. The first physical NIC name of each Harvester node always defaults to eth0. It is recommended to choose a separate NIC for the VLAN other than the one used for the management network(the one selected during the Harvester installation) for better network performance and isolation. (Note: modify the default VLAN network setting will not change the existing configured host networks).
+
+  ![image-20210408233858097](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210408233858097.png)
+
+- (optional) Users can always customize each node's VLAN network configuration via going to **HOST > Network** tab.
```
Comment (guangbochen) [inline docs/harvester-network.md]: \s\HOST\the Host

File: docs/harvester-network.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+  ![image-20210408235822223](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210408235822223.png)
+
+- Creating a new VLAN network via going to **Advanced>Networks** page and clicking the **Create** button.
```
Comment (guangbochen) [inline docs/harvester-network.md]: \s\Advanced\the Advanced

**Review by guangbochen — APPROVED** [review]
LGTM

File: docs/harvester-network.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+Harvester adopts [flannel](https://github.com/flannel-io/flannel) as the default CNI to implement the management network. It's an internal network, which means the user can only access the VM's management network within its cluster nodes or pods.
+
+#### VLAN
+
+[Harvester network-controller](https://github.com/rancher/harvester-network-controller) leverages the [multus](https://github.com/k8snetworkplumbingwg/multus-cni) and [bridge](https://www.cni.dev/plugins/current/main/bridge/) CNI plugins to implement the VLAN.  
+
+Below is a use case of the VLAN in Harvester.
+
+![image-20210412181210146](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210412181210146.png)
```
Comment (gitlawr) [inline docs/harvester-network.md]: Can we put all image files to https://github.com/rancher/harvester/tree/master/docs/assets

File: docs/harvester-network.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+- Harvester network controller use a bridge for a node and a pair of veth for a VM to implement the VLAN. The bridge acts as a switch to forward the network traffic from or to VMs and the veth pair is like the connected ports between vms and switch.
```
Comment (gitlawr) [inline docs/harvester-network.md]: s/use/uses/

File: docs/harvester-network.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+  ![image-20210408235822223](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210408235822223.png)
+
+- Creating a new VLAN network via going to the **Advanced>Networks** page and clicking the **Create** button.
+
+  ![image-20210408232931111](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210408232931111-20210409000151223.png)
+
+- Creating a VM and add the network configurations.
+
+  - Only the first network card will be enabled by default, the user can either choose to use a management network or VLAN network. (Note: will need to select `Install guest agent` option in the `Advanced Options` tab to get the VLAN network IP address from the Harvester UI)
```
Comment (gitlawr) [inline docs/harvester-network.md]: s/network card/network interface card/
s/, the user/. Users/

File: docs/harvester-network.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+
+    ![image-20210412175826097](https://raw.githubusercontent.com/yaocw2020/img/master/image-20210412175826097.png)
+
+  - Users can choose to add one or multiple network cards, the additional network card configurations can be set via cloud-init network data. e.g.
```
Comment (gitlawr) [inline docs/harvester-network.md]: Users can choose to add one or multiple network interface cards. Additional network interface card configurations can be set via cloud-init network data.

File: docs/harvester-network.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+    ```
+    version: 1
+    config:
+      - type: physical
+        name: enp1s0 # name is various upon OS image
```
Comment (gitlawr) [inline docs/harvester-network.md]: s/is various upon/varies upon/

**Comment by gitlawr** [comment]
Can we clarify the usage of VLAN 1 given that it's the default PVID in most switches?

**Comment by yaocw2020** [comment]
> Can we clarify the usage of VLAN 1 given that it's the default PVID in most switches?

Add "- Users can use VLAN with `PVID`(default 1) to communicate with any normal untagged traffic.". And I have submitted a [issue to add VLAN 1 by default](https://github.com/rancher/harvester/issues/704).

**Review by gitlawr — APPROVED** [review]
LGTM

**Pull request merged**